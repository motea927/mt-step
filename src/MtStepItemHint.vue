<template>
  <div
    :class="[
      'mt-step-item-hint',
      `mt-step-item-hint--${props.hint?.position ?? 'top'}`,
    ]"
    :style="positionStyle"
  >
    {{ hint?.text }}
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue'
import { useRect } from './composables/useRect'

const props = defineProps<{
  selector: string
  hint?: {
    text: string
    position?: 'top' | 'right' | 'bottom' | 'left'
  }
}>()

const { rect } = useRect(props.selector)

const positionStyle = computed(() => {
  const { top, left, width, height } = rect.value
  const position = props.hint?.position ?? 'top'
  const styleMap = {
    top: {
      top: `${top}px`,
      left: `${left + width / 2}px`,
    },
    right: {
      top: `${top + height / 2}px`,
      left: `${left + width}px`,
    },
    bottom: {
      top: `${top + height}px`,
      left: `${left + width / 2}px`,
    },
    left: {
      top: `${top + height / 2}px`,
      left: `${left}px`,
    },
  }

  return styleMap[position]
})
</script>

<style lang="scss">
.mt-step-item-hint {
  position: absolute;
  background-color: #fff;
  padding: 8px;
  border-radius: 4px;
  z-index: var(--z-index-hint);
  color: #666;

  &::after {
    content: '';
    position: absolute;
    border: 5px solid #fff;
  }

  &--top {
    transform: translate(-50%, calc(-100% - 20px));
  }

  &--top::after {
    border-color: #fff transparent transparent;
    bottom: -10px;
    left: 50%;
    transform: translateX(-50%);
  }

  &--right {
    transform: translate(20px, calc(-50%));
  }

  &--right::after {
    border-color: transparent #fff transparent transparent;
    top: 50%;
    left: -10px;
    transform: translateY(-50%);
  }

  &--bottom {
    transform: translate(-50%, 20px);
  }

  &--bottom::after {
    border-color: transparent transparent #fff;
    top: -10px;
    left: 50%;
    transform: translateX(-50%);
  }

  &--left {
    transform: translate(calc(-100% - 20px), -50%);
  }

  &--left::after {
    border-color: transparent transparent transparent #fff;
    top: 50%;
    right: -10px;
    transform: translateY(-50%);
  }
}
</style>
