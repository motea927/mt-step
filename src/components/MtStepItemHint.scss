.mt-step-item-hint {
  position: absolute;
  z-index: var(--z-index-hint);

  &--default {
    background-color: #fff;
    padding: 8px;
    border-radius: 4px;
    color: #666;
  }

  &--default::after {
    content: '';
    position: absolute;
    border: 5px solid #fff;
  }

  &--top {
    transform: translate(-50%, calc(-100% - 20px));
  }

  &--top::after {
    border-color: #fff transparent transparent;
    bottom: -10px;
    left: 50%;
    transform: translateX(-50%);
  }

  &--right {
    transform: translate(20px, calc(-50%));
  }

  &--right::after {
    border-color: transparent #fff transparent transparent;
    top: 50%;
    left: -10px;
    transform: translateY(-50%);
  }

  &--bottom {
    transform: translate(-50%, 20px);
  }

  &--bottom::after {
    border-color: transparent transparent #fff;
    top: -10px;
    left: 50%;
    transform: translateX(-50%);
  }

  &--left {
    transform: translate(calc(-100% - 20px), -50%);
  }

  &--left::after {
    border-color: transparent transparent transparent #fff;
    top: 50%;
    right: -10px;
    transform: translateY(-50%);
  }
}
