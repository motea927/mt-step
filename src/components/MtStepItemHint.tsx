import { defineComponent, computed, h } from 'vue-demi'
import { useRect } from '@/composables/useRect'
import type { PropType } from 'vue-demi'

import './MtStepItemHint.scss'

const MtStepItemHint = defineComponent({
  props: {
    selector: { type: String as PropType<string>, required: true },
    hint: {
      type: Object as PropType<{
        text: string
        position?: 'top' | 'right' | 'bottom' | 'left'
      }>,
      required: true,
    },
  },

  setup(props) {
    const { rect } = useRect(props.selector)
    const positionStyle = computed(() => {
      const { top, left, width, height } = rect.value
      const position = props.hint?.position ?? 'top'
      const styleMap = {
        top: {
          top: `${top}px`,
          left: `${left + width / 2}px`,
        },
        right: {
          top: `${top + height / 2}px`,
          left: `${left + width}px`,
        },
        bottom: {
          top: `${top + height}px`,
          left: `${left + width / 2}px`,
        },
        left: {
          top: `${top + height / 2}px`,
          left: `${left}px`,
        },
      }

      return styleMap[position]
    })

    return () =>
      h(
        'div',
        {
          class: [
            'mt-step-item-hint',
            `mt-step-item-hint--${props.hint.position ?? 'top'}`,
          ],
          style: positionStyle.value,
        },
        props.hint.text
      )
  },
})

export default MtStepItemHint
