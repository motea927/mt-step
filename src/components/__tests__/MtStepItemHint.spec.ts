import { describe, beforeEach, afterEach, it, expect, test } from 'vitest'
import { mount } from '@vue/test-utils'
import MtStepItemHint from '@/components/MtStepItemHint'

const mountElement = (hintPosition?: 'top' | 'right' | 'bottom' | 'left') =>
  mount(MtStepItemHint, {
    props: {
      selector: '#test',
      hint: {
        text: 'something',
        ...(hintPosition && { position: hintPosition }),
      },
    },
  })

describe('MtStepItemHint', () => {
  beforeEach(() => {
    // Create a target element for testing
    const targetElement = document.createElement('div')
    targetElement.id = 'test'
    document.body.appendChild(targetElement)
    targetElement.style.position = ''
  })

  afterEach(() => {
    // Remove the target element after testing
    ;['#test'].forEach((selector) => {
      const targetElement = document.querySelector(selector)
      document.body.removeChild(targetElement as Element)
    })
  })

  it('Render correct hint text', () => {
    const wrapper = mountElement()
    expect(wrapper.html()).toContain('something')
  })

  it('props.hint.position is optional, default is top', () => {
    const wrapper = mountElement()
    expect(wrapper.html()).toContain('mt-step-item-hint--top')
    expect(wrapper.element).toMatchSnapshot()
  })

  test.each(['top', 'right', 'bottom', 'left'] as const)(
    'render correct className with props.hint.position %s',
    (position) => {
      const wrapper = mountElement(position)
      expect(wrapper.html()).toContain(position)
      expect(wrapper.element).toMatchSnapshot()
    }
  )
})
