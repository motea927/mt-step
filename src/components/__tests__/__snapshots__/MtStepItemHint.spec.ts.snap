// Vitest Snapshot v1

exports[`MtStepItemHint > Default position is top 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--top"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > Renders correctly 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--top"
  style="top: 0px; left: 0px;"
/>
`;

exports[`MtStepItemHint > position props is optional, default is top 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--top"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > props.hint.position is optional, default is top 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--top"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--top"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position 2`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--right"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position 3`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--bottom"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position 4`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--left"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position bottom 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--bottom"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position left 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--left"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position right 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--right"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;

exports[`MtStepItemHint > render correct className with props.hint.position top 1`] = `
<div
  class="mt-step-item-hint mt-step-item-hint--top"
  style="top: 0px; left: 0px;"
>
  something
</div>
`;
